### ISSUE TITLE
nbextensions should be installed and enabled on a per-kernel basis, not per-user.

### ISSUE THREAD

**Issue by ksze** [jupyter/notebook#2841]
We are running into some problems with the way the jupyter notebook extension system works.

It appears that notebook extensions are installed and enabled either system-wide, or per-user.

This makes life difficult when sharing notebooks with other users on JupyterHub. The other user would need access to the terminal install and enable any extensions needed by a given notebook. God forbid if different kernels need different versions of the same extension.

I believe that installing and enabling notebook extensions on a per-kernel basis would make more sense. If you are going to share a notebook to another user, chances are that you want the other use to run it with the same kernel.

**Comment by takluyver** [comment]
Many (most?) extensions do not depend on the kernel, so it wouldn't make sense to install them per-kernel. Part of the point of nbextensions is to provide functionality that works across different kernels.

Kernels get a `kernel.js` file which is loaded in the frontend when the kernel starts. Customisation can be done there. See e.g. [IRkernel's kernel.js](https://github.com/IRkernel/IRkernel/blob/master/inst/kernelspec/kernel.js). You can also put files in the nbextensions directory but not configure them as an nbextension, and then use a `require()` call to load them - either from `kernel.js` or from Javascript output produced by running code. [Mobilechelonian](https://github.com/takluyver/mobilechelonian) uses this technique, for example.

If you tell us a bit more about what you're trying to achieve, we may be able to point you to more ideas.

**Comment by ksze** [comment]
A prime example is widgetsnbextension from ipywidgets, which, ironically, is an "official" jupyter notebook extension. widgetsnbextension requires both JavaScript and Python module to work, but if your notebook server loads the JavaScript for widgetsnbextension version 2.0.0 while your kernel has the widgetsnbextension version 3.0.0 Python package, it won't work: the notebook would complain about incompatibility.

But why should I expect that every kernel has the same widgetsnbextension version 2.0.0 Python package? It's entirely possible that different notebooks depend on different kernels, which are based on different virtualenvs, which have different versions of the widgetsnbextension Python package.

If I understand correctly, the current mechanism installs and enables extensions either per-user (`--user`) or per-virtualenv (`--sys-prefix`). But even in the per-virtualenv case, extension JavaScript is loaded from the virtualenv that is used to run the notebook _server_ (this includes `jupyter notebook` and `jupyterhub-singleuser` instances spawned by JupyterHub). Extension JavaScript is not loaded from the virtualenv of the notebook's kernel. This mechanism seems really broken to me.

**Comment by ksze** [comment]
I'm also not sure how putting `require()` in kernel.js helps. In order for that to work, the notebook server would need to be able to serve static files from the kernel, but that appears to have been an ongoing issue since 2015 (#320).

For the moment, it appears that the best workaround is to read the JavaScript file content in the kernel and serve it up via `IPython.display.Javascript`. It's trivial for the python module to find its own static files (through a combination of `os.path.join()` and `__file__`). It may be relatively inefficient, but it saves me the trouble of trying to match JavaScript and Python module versions among the notebook server and the kernels.

**Comment by mlucool** [comment]
@ksze did you ever fully solve this?

**Comment by marvinxu-free** [comment]
if i have virtualenv for python2 and python3, how can i install nbextensions for each virtualenv?

**Comment by rolweber** [comment]
@marvinxu-free: The notebook extension must be installed in the environment from which you run Jupyter. If you want to run it from either environment, install it twice. Switch to one virtualenv and follow the installation instructions of the notebook extension. Then switch to the other virtualenv and repeat.

If you want to run Jupyter only from one envirnment, but you have kernels defined from both environments, you still need to install the notebook extensions only once, in the environment from which you run Jupyter.

Sometimes, this is complicated by packaging. For example, widgetsnbextensions is a notebook extension that needs to be installed only once. ipywidgets are a Python package that needs to be installed for each kernel. But because ipywidgets defines a dependency on widgetsnbextensions, you're likely to have the extensions installed twice, too. Which doesn't hurt, by the way.

**Comment by ksze** [comment]
@mlucool No. I never got to the bottom of it. For now I just let nbextension have its way and install in both the notebook kernel and the jupyter server's virtualenv.

**Comment by jasongrout** [comment]
> Sometimes, this is complicated by packaging. For example, widgetsnbextensions is a notebook extension that needs to be installed only once. ipywidgets are a Python package that needs to be installed for each kernel. But because ipywidgets defines a dependency on widgetsnbextensions, you're likely to have the extensions installed twice, too. Which doesn't hurt, by the way.

FYI, there are conversations about breaking up this dependency relationship: https://github.com/jupyter-widgets/ipywidgets/issues/1892
